File: fusion_core/python_prototype/tests/test_ukf.py
"""MIP Section 11.2: feed synthetic sensor data with known ground
truth (straight line, constant turn) and check the UKF converges to
the expected state.

Run with: python -m pytest fusion_core/python_prototype/tests/ -v
(from repo root, or inside `docker compose run --rm ml bash`).
"""

from __future__ import annotations

import numpy as np
import pytest

from fusion_core.python_prototype.ukf import DualChannelUkf, FusionConfig, UkfState


def _synthetic_truth(kind: str, duration_s: float, dt_s: float, speed_mps: float, turn_rate_dps: float = 0.0):
    n = int(round(duration_s / dt_s)) + 1
    t = np.arange(n) * dt_s
    turn_rate_rps = np.deg2rad(turn_rate_dps) if kind == "constant_turn" else 0.0
    heading = turn_rate_rps * t
    vel = np.stack([speed_mps * np.cos(heading), speed_mps * np.sin(heading)], axis=1)
    pos = np.zeros((n, 2))
    pos[1:] = np.cumsum((vel[:-1] + vel[1:]) / 2 * dt_s, axis=0)
    return t, pos, vel, heading, np.full(n, turn_rate_rps)


def _run_filter(t, pos, vel, heading, gyro_yaw_true, rng, gnss_every_n=1, noise_std=0.05):
    dt = float(t[1] - t[0])
    initial = UkfState(pos=pos[0].copy(), vel=vel[0].copy(), heading=float(heading[0]))
    ukf = DualChannelUkf(initial, FusionConfig())

    est_pos = np.zeros_like(pos)
    est_pos[0] = pos[0]

    for i in range(1, len(t)):
        speed_true = float(np.linalg.norm(vel[i]))
        channel_a = speed_true + rng.normal(0.0, noise_std)
        channel_b = speed_true + rng.normal(0.0, noise_std * 3)
        gyro_meas = gyro_yaw_true[i - 1] + rng.normal(0.0, 0.005)

        gnss_available = (i % gnss_every_n) == 0
        gnss_pos = pos[i].copy() + rng.normal(0.0, 0.5, size=2) if gnss_available else None
        gnss_vel = vel[i].copy() + rng.normal(0.0, 0.05, size=2) if gnss_available else None

        state = ukf.step(
            dt=dt,
            gyro_yaw=gyro_meas,
            channel_a_speed=channel_a,
            channel_b_speed=channel_b,
            gnss_pos=gnss_pos,
            gnss_vel=gnss_vel,
            road_signature_pos=None,
            road_signature_confidence=0.0,
        )
        est_pos[i] = state.pos

    return est_pos


@pytest.mark.parametrize("kind,turn_rate_dps", [("straight", 0.0), ("constant_turn", 3.0)])
def test_ukf_converges_with_gnss(kind, turn_rate_dps):
    """With GNSS continuously available, the fused position should
    stay tightly bound to ground truth throughout - this is the
    baseline sanity check before testing blackout behavior."""
    t, pos, vel, heading, gyro_yaw_true = _synthetic_truth(
        kind, duration_s=60.0, dt_s=0.1, speed_mps=16.7, turn_rate_dps=turn_rate_dps
    )
    rng = np.random.default_rng(0)
    est_pos = _run_filter(t, pos, vel, heading, gyro_yaw_true, rng, gnss_every_n=1)

    # Skip the first couple of seconds to let the filter converge from
    # its initial covariance before checking steady-state error.
    settle_idx = int(2.0 / (t[1] - t[0]))
    err = np.linalg.norm(est_pos[settle_idx:] - pos[settle_idx:], axis=1)
    assert np.mean(err) < 3.0, f"mean position error too high with GNSS available: {np.mean(err):.2f} m"


@pytest.mark.parametrize("kind,turn_rate_dps", [("straight", 0.0), ("constant_turn", 3.0)])
def test_ukf_bounds_drift_during_blackout(kind, turn_rate_dps):
    """During a GNSS blackout, the filter (Channel A/B only) should
    stay far closer to ground truth than naive open-loop double
    integration would (HLD Section 1) - it won't be perfect, but it
    must not blow up."""
    duration_s = 120.0
    dt_s = 0.1
    t, pos, vel, heading, gyro_yaw_true = _synthetic_truth(
        kind, duration_s=duration_s, dt_s=dt_s, speed_mps=16.7, turn_rate_dps=turn_rate_dps
    )
    rng = np.random.default_rng(1)

    initial = UkfState(pos=pos[0].copy(), vel=vel[0].copy(), heading=float(heading[0]))
    ukf = DualChannelUkf(initial, FusionConfig())

    blackout_start_s, blackout_end_s = 40.0, 100.0
    est_pos = np.zeros_like(pos)
    est_pos[0] = pos[0]

    for i in range(1, len(t)):
        speed_true = float(np.linalg.norm(vel[i]))
        channel_a = speed_true + rng.normal(0.0, 0.3)
        channel_b = speed_true + rng.normal(0.0, 1.0)
        gyro_meas = gyro_yaw_true[i - 1] + rng.normal(0.0, 0.01)

        in_blackout = blackout_start_s <= t[i] < blackout_end_s
        gnss_pos = None if in_blackout else pos[i].copy() + rng.normal(0.0, 0.5, size=2)
        gnss_vel = None if in_blackout else vel[i].copy() + rng.normal(0.0, 0.05, size=2)

        state = ukf.step(
            dt=dt_s,
            gyro_yaw=gyro_meas,
            channel_a_speed=channel_a,
            channel_b_speed=channel_b,
            gnss_pos=gnss_pos,
            gnss_vel=gnss_vel,
            road_signature_pos=None,
            road_signature_confidence=0.0,
        )
        est_pos[i] = state.pos

    end_idx = np.searchsorted(t, blackout_end_s) - 1
    error_at_end_m = float(np.linalg.norm(est_pos[end_idx] - pos[end_idx]))
    distance_traveled_m = float(
        np.sum(np.linalg.norm(np.diff(pos[np.searchsorted(t, blackout_start_s):end_idx + 1], axis=0), axis=1))
    )
    drift_pct = error_at_end_m / distance_traveled_m * 100.0

    # Loose bound for a synthetic-noise, no-road-signature-anchor
    # scenario - not asserting the <10% product benchmark itself here
    # (that needs the real Channel A/B and the anchor), just that a
    # correctly-wired UKF meaningfully beats raw open-loop drift.
    assert drift_pct < 15.0, f"drift too high during blackout: {drift_pct:.2f}%"


def test_road_signature_anchor_pulls_position_toward_segment():
    """A high-confidence road-signature match should measurably pull
    the fused position toward the anchor, per Section 5.3 point 4."""
    initial = UkfState(pos=np.array([0.0, 0.0]), vel=np.array([16.7, 0.0]), heading=0.0)
    ukf = DualChannelUkf(initial, FusionConfig())

    # Run a few blackout steps with no anchor, drifting slightly.
    for _ in range(20):
        ukf.step(
            dt=0.1,
            gyro_yaw=0.02,  # small unmodeled turn the filter can't fully explain from Ch A/B alone
            channel_a_speed=16.7,
            channel_b_speed=16.7,
            gnss_pos=None,
            gnss_vel=None,
            road_signature_pos=None,
            road_signature_confidence=0.0,
        )

    pos_before = ukf.ukf.x[:2].copy()
    anchor = pos_before + np.array([5.0, -5.0])  # a deliberately offset "true" segment midpoint

    state = ukf.step(
        dt=0.1,
        gyro_yaw=0.02,
        channel_a_speed=16.7,
        channel_b_speed=16.7,
        gnss_pos=None,
        gnss_vel=None,
        road_signature_pos=anchor,
        road_signature_confidence=0.9,
    )

    dist_before = float(np.linalg.norm(pos_before - anchor))
    dist_after = float(np.linalg.norm(state.pos - anchor))
    assert dist_after < dist_before, "road-signature anchor did not pull position toward the segment"


def test_channel_a_downweighted_on_disagreement():
    """Section 5.4: when Channel A and Channel B disagree sharply, the
    filter should lean toward Channel B rather than trusting Channel A
    outright - verified indirectly by checking the filter does not
    fully adopt Channel A's (wrong) speed in one step."""
    initial = UkfState(pos=np.array([0.0, 0.0]), vel=np.array([16.7, 0.0]), heading=0.0)
    ukf = DualChannelUkf(initial, FusionConfig())

    # Channel A reports a wild, wrong speed (simulating a pothole
    # gyro-glitch/mount-slip event); Channel B stays near truth.
    state = ukf.step(
        dt=0.1,
        gyro_yaw=0.0,
        channel_a_speed=40.0,
        channel_b_speed=16.7,
        gnss_pos=None,
        gnss_vel=None,
        road_signature_pos=None,
        road_signature_confidence=0.0,
    )
    fused_speed = float(np.linalg.norm(state.vel))
    # Should sit far closer to Channel B's 16.7 than to Channel A's 40.
    assert abs(fused_speed - 16.7) < abs(fused_speed - 40.0)


def test_nhc_disabled_by_default():
    """Regression guard for a real finding (see FusionConfig.enable_nhc's
    docstring): enabling the NHC pseudo-measurement measurably hurts
    drift on this system's synthetic constant-turn benchmark, because
    Channel A/B's own MIP-specified design (rotating scalar speed by
    the current heading estimate) already bakes in the same
    zero-lateral-velocity assumption NHC asserts a second time. If
    this test ever fails because someone flipped the default to
    True, that's a deliberate design change, not a typo - it should
    come with a fresh ablation re-run (see this file's git history for
    the numbers) and a docstring update, not a silent flip."""
    assert FusionConfig().enable_nhc is False

File: fusion_core/cpp/tests/test_ukf.cpp
// MIP Section 11.2: feed synthetic sensor data with known ground
// truth (straight line, constant turn) and check the UKF converges to
// the expected state. Mirrors
// fusion_core/python_prototype/tests/test_ukf.py's four test cases -
// keep the two in sync if either changes.

#include <cmath>
#include <random>
#include <vector>

#include <gtest/gtest.h>

#include "fusion_core/ukf.hpp"

using fusion_core::DualChannelUkf;
using fusion_core::FusionConfig;
using fusion_core::UkfState;

namespace {

struct SyntheticTruth {
  std::vector<double> t;
  std::vector<Eigen::Vector2d> pos;
  std::vector<Eigen::Vector2d> vel;
  std::vector<double> heading;
  std::vector<double> gyroYawTrue;
};

SyntheticTruth generateSyntheticTruth(double durationS, double dtS, double speedMps,
                                       double turnRateDps) {
  const int n = static_cast<int>(std::round(durationS / dtS)) + 1;
  const double turnRateRps = turnRateDps * M_PI / 180.0;

  SyntheticTruth truth;
  truth.t.resize(n);
  truth.pos.resize(n);
  truth.vel.resize(n);
  truth.heading.resize(n);
  truth.gyroYawTrue.resize(n, turnRateRps);

  truth.pos[0] = Eigen::Vector2d(0.0, 0.0);
  for (int i = 0; i < n; ++i) {
    truth.t[i] = i * dtS;
    truth.heading[i] = turnRateRps * truth.t[i];
    truth.vel[i] = Eigen::Vector2d(speedMps * std::cos(truth.heading[i]),
                                    speedMps * std::sin(truth.heading[i]));
  }
  for (int i = 1; i < n; ++i) {
    truth.pos[i] = truth.pos[i - 1] + (truth.vel[i - 1] + truth.vel[i]) / 2.0 * dtS;
  }
  return truth;
}

double pathLength(const std::vector<Eigen::Vector2d>& pos, int start, int end) {
  double total = 0.0;
  for (int i = start + 1; i <= end; ++i) {
    total += (pos[i] - pos[i - 1]).norm();
  }
  return total;
}

}  // namespace

class UkfConvergenceTest : public ::testing::TestWithParam<std::pair<std::string, double>> {};

TEST_P(UkfConvergenceTest, ConvergesWithGnssAvailable) {
  const double turnRateDps = GetParam().second;
  const double durationS = 60.0;
  const double dtS = 0.1;
  const double speedMps = 16.7;

  SyntheticTruth truth = generateSyntheticTruth(durationS, dtS, speedMps, turnRateDps);
  const int n = static_cast<int>(truth.t.size());

  std::mt19937 rng(0);
  std::normal_distribution<double> speedNoise(0.0, 0.05);
  std::normal_distribution<double> gyroNoise(0.0, 0.005);
  std::normal_distribution<double> gnssPosNoise(0.0, 0.5);
  std::normal_distribution<double> gnssVelNoise(0.0, 0.05);

  UkfState initial{truth.pos[0], truth.vel[0], truth.heading[0]};
  DualChannelUkf ukf(initial, FusionConfig());

  std::vector<Eigen::Vector2d> estPos(n);
  estPos[0] = truth.pos[0];

  for (int i = 1; i < n; ++i) {
    const double speedTrue = truth.vel[i].norm();
    const double channelA = speedTrue + speedNoise(rng);
    const double channelB = speedTrue + speedNoise(rng) * 3.0;
    const double gyroMeas = truth.gyroYawTrue[i - 1] + gyroNoise(rng);

    Eigen::Vector2d gnssPos =
        truth.pos[i] + Eigen::Vector2d(gnssPosNoise(rng), gnssPosNoise(rng));
    Eigen::Vector2d gnssVel =
        truth.vel[i] + Eigen::Vector2d(gnssVelNoise(rng), gnssVelNoise(rng));

    UkfState state = ukf.step(dtS, gyroMeas, channelA, channelB, gnssPos, gnssVel, std::nullopt, 0.0);
    estPos[i] = state.pos;
  }

  const int settleIdx = static_cast<int>(2.0 / dtS);
  double sumErr = 0.0;
  for (int i = settleIdx; i < n; ++i) {
    sumErr += (estPos[i] - truth.pos[i]).norm();
  }
  const double meanErr = sumErr / (n - settleIdx);
  EXPECT_LT(meanErr, 3.0) << "mean position error too high with GNSS available: " << meanErr << " m";
}

TEST_P(UkfConvergenceTest, BoundsDriftDuringBlackout) {
  const double turnRateDps = GetParam().second;
  const double durationS = 120.0;
  const double dtS = 0.1;
  const double speedMps = 16.7;
  const double blackoutStartS = 40.0;
  const double blackoutEndS = 100.0;

  SyntheticTruth truth = generateSyntheticTruth(durationS, dtS, speedMps, turnRateDps);
  const int n = static_cast<int>(truth.t.size());

  std::mt19937 rng(1);
  std::normal_distribution<double> aNoise(0.0, 0.3);
  std::normal_distribution<double> bNoise(0.0, 1.0);
  std::normal_distribution<double> gyroNoise(0.0, 0.01);
  std::normal_distribution<double> gnssPosNoise(0.0, 0.5);
  std::normal_distribution<double> gnssVelNoise(0.0, 0.05);

  UkfState initial{truth.pos[0], truth.vel[0], truth.heading[0]};
  DualChannelUkf ukf(initial, FusionConfig());

  std::vector<Eigen::Vector2d> estPos(n);
  estPos[0] = truth.pos[0];

  int endIdx = 0;
  for (int i = 1; i < n; ++i) {
    const double speedTrue = truth.vel[i].norm();
    const double channelA = speedTrue + aNoise(rng);
    const double channelB = speedTrue + bNoise(rng);
    const double gyroMeas = truth.gyroYawTrue[i - 1] + gyroNoise(rng);

    const bool inBlackout = truth.t[i] >= blackoutStartS && truth.t[i] < blackoutEndS;

    std::optional<Eigen::Vector2d> gnssPos;
    std::optional<Eigen::Vector2d> gnssVel;
    if (!inBlackout) {
      gnssPos = truth.pos[i] + Eigen::Vector2d(gnssPosNoise(rng), gnssPosNoise(rng));
      gnssVel = truth.vel[i] + Eigen::Vector2d(gnssVelNoise(rng), gnssVelNoise(rng));
    } else {
      endIdx = i;
    }

    UkfState state = ukf.step(dtS, gyroMeas, channelA, channelB, gnssPos, gnssVel, std::nullopt, 0.0);
    estPos[i] = state.pos;
  }

  const double errorAtEndM = (estPos[endIdx] - truth.pos[endIdx]).norm();
  int startIdx = 0;
  for (int i = 0; i < n; ++i) {
    if (truth.t[i] >= blackoutStartS) {
      startIdx = i;
      break;
    }
  }
  const double distanceTraveledM = pathLength(truth.pos, startIdx, endIdx);
  const double driftPct = errorAtEndM / distanceTraveledM * 100.0;

  // Loose bound mirroring the Python test - not asserting the <10%
  // product benchmark itself here (needs real Channel A/B and the
  // road-signature anchor), just that a correctly-ported UKF
  // meaningfully beats raw open-loop drift.
  EXPECT_LT(driftPct, 15.0) << "drift too high during blackout: " << driftPct << "%";
}

INSTANTIATE_TEST_SUITE_P(StraightAndConstantTurn, UkfConvergenceTest,
                          ::testing::Values(std::make_pair(std::string("straight"), 0.0),
                                             std::make_pair(std::string("constant_turn"), 3.0)));

TEST(UkfRoadSignatureTest, AnchorPullsPositionTowardSegment) {
  UkfState initial{Eigen::Vector2d(0.0, 0.0), Eigen::Vector2d(16.7, 0.0), 0.0};
  DualChannelUkf ukf(initial, FusionConfig());

  // Run a few blackout steps with no anchor, drifting slightly.
  for (int i = 0; i < 20; ++i) {
    ukf.step(0.1, 0.02, 16.7, 16.7, std::nullopt, std::nullopt, std::nullopt, 0.0);
  }

  Eigen::Vector2d posBefore = ukf.rawState().head<2>();
  Eigen::Vector2d anchor = posBefore + Eigen::Vector2d(5.0, -5.0);

  UkfState state = ukf.step(0.1, 0.02, 16.7, 16.7, std::nullopt, std::nullopt, anchor, 0.9);

  const double distBefore = (posBefore - anchor).norm();
  const double distAfter = (state.pos - anchor).norm();
  EXPECT_LT(distAfter, distBefore) << "road-signature anchor did not pull position toward the segment";
}

TEST(UkfTrustWeightingTest, ChannelADownweightedOnDisagreement) {
  UkfState initial{Eigen::Vector2d(0.0, 0.0), Eigen::Vector2d(16.7, 0.0), 0.0};
  DualChannelUkf ukf(initial, FusionConfig());

  // Channel A reports a wild, wrong speed (simulating a pothole
  // gyro-glitch/mount-slip event); Channel B stays near truth.
  UkfState state = ukf.step(0.1, 0.0, 40.0, 16.7, std::nullopt, std::nullopt, std::nullopt, 0.0);
  const double fusedSpeed = state.vel.norm();
  EXPECT_LT(std::abs(fusedSpeed - 16.7), std::abs(fusedSpeed - 40.0));
}

TEST(UkfNhcTest, DisabledByDefault) {
  // Regression guard for a real finding (see FusionConfig::enable_nhc's
  // docstring, ukf.hpp): enabling NHC measurably hurts drift on this
  // system's synthetic constant-turn benchmark, because Channel A/B's
  // own MIP-specified design already bakes in the same
  // zero-lateral-velocity assumption NHC asserts a second time. If
  // this ever fails because someone flipped the default to true,
  // that's a deliberate design change - it should come with a fresh
  // ablation re-run and a docstring update, not a silent flip.
  EXPECT_FALSE(FusionConfig().enable_nhc);
}
